### PR TITLE
[SPARK-8280][SPARK-8281][SQL]Handle NaN, null and Infinity in math

### PR DESCRIPTION
**Pull request by yjshen**
Just as their counterparts in Hive:

`UDFAsin` & `UDFAcos` would just return NaN if its argument is NaN or its absolute value is greater than 1, therefore I think it's enough to just remove NaN checking before return.

`Log*` would first check `argument <= 0.0` and return null for 0.0 instead of `-Infinity`

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, ok to test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 1]
```diff
@@ -81,9 +80,31 @@ abstract class UnaryMathExpression(f: Double => Double, name: String)
...
+      null
+    } else {
+      f(evalE.asInstanceOf[Double])
+    }
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val eval = child.gen(ctx)
+    eval.code + s"""
+      boolean ${ev.isNull} = ${eval.isNull} || Double.valueOf(${eval.primitive}) <= 0.0;
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: isn't this just

```
${eval.primitive} <= 0.0
```

instead of calling Double.valueOf?

**Comment by rxin** [comment]
@yijieshen it'd make sense to merge this after https://github.com/apache/spark/pull/6725

It might change your code here.

**Comment by rxin** [comment]
btw I was thinking maybe Log2 and the other log functions can just be child class of the Logarithm class added in #6725?

**Comment by yijieshen** [comment]
I've merged #6725 in the latest commit.

Also tried to refactor all the log functions to subclass the Logarithm class added there, but find wried  and hard to understand then, so just discard that version.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 2]
```diff
@@ -248,30 +269,81 @@ case class Hypot(left: Expression, right: Expression)
...
+        if (${eval2.isNull} || ${eval2.primitive} <= 0.0) {
+          ${ev.isNull} = true;
+        } else {
+          ${ev.primitive} = java.lang.Math.${funcName}(${eval2.primitive}) /
+           java.lang.Math.${funcName}(${eval1.primitive});
+        }
       }
     """
   }
+
```
Comment (yijieshen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Hive's UDFLog doesn't support base in range (0.0, 1.0], it would just eval to null in this case, just act as the commented out code below.

I'm not sure it we also want to behave  like this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 3]
```diff
@@ -248,30 +269,81 @@ case class Hypot(left: Expression, right: Expression)
...
+//    if (evalE1 == null || evalE2 == null) {
+//      null
+//    } else {
+//      if (evalE1.asInstanceOf[Double] <= 1.0 || evalE2.asInstanceOf[Double] <= 0.0) {
+//        null
+//      } else {
+//        math.log(evalE2.asInstanceOf[Double]) / math.log(evalE1.asInstanceOf[Double])
+//      }
+//    }
+//  }
```
Comment (yijieshen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: Hive's UDFLog doesn't support base in range (0.0, 1.0], it would just eval to null in this case, just act as the commented out code below.

I'm not sure it we also want to behave like this.

**Comment by yijieshen** [comment]
I think we should discuss the preferred behaviour of `binary log`

**Comment by yijieshen** [comment]
Seems build failure is not related to this PR

**Comment by yijieshen** [comment]
retest this please.

**Comment by JoshRosen** [comment]
The MiMa test failure is due to an artifact missing from a Maven staging repository; I have opened #6974 to hotfix.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala [hunk 4]
```diff
@@ -259,19 +285,14 @@ case class Atan2(left: Expression, right: Expression)
         null
       } else {
         // With codegen, the values returned by -0.0 and 0.0 are different. Handled with +0.0
-        val result = math.atan2(evalE1.asInstanceOf[Double] + 0.0,
+        math.atan2(evalE1.asInstanceOf[Double] + 0.0,
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: i don't think u need to wrap here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala]: we should support these. just remove the commented code, and add to inline comment for log that we support (0.0, 1.0], unlike hive.

File: sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala [hunk 5]
```diff
@@ -69,7 +69,7 @@ class MathExpressionsSuite extends QueryTest {
     if (f(-1) === math.log1p(-1)) {
       checkAnswer(
         nnDoubleData.select(c('b)),
-        (1 to 9).map(n => Row(f(n * -0.1))) :+ Row(Double.NegativeInfinity)
+        (1 to 9).map(n => Row(f(n * -0.1))) :+ Row(null)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/MathExpressionsSuite.scala]: why the change from -inf to null?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 6]
```diff
@@ -92,6 +105,62 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+      case e: Exception => fail(s"Exception evaluating $expression", e)
+    }
+    if (!actual.asInstanceOf[Double].isNaN) {
+      val input = if (inputRow == EmptyRow) "" else s", input: $inputRow"
+      fail(s"Incorrect evaluation (codegen off): $expression, " +
+        s"actual: $actual, " +
+        s"expected: NaN$input")
+    }
+  }
+
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: remove the extra blank line here

**Comment by rxin** [comment]
@yijieshen can you bring this up to date?

**Comment by yijieshen** [comment]
@rxin, do you mind my close this one and create a new pr for this? The base code is quite outdated.

**Comment by rxin** [comment]
Not at all. Feel free to close this one and submit a new one.
.

**Comment by yijieshen** [comment]
ok, will do soon

**Comment by yijieshen** [comment]
close this and continue in #7451
